v2/api_error.go

```diff
@@ -0,0 +1,185 @@
+			f = append(f, x.Field)
+			desc = append(desc, x.Description)
+		}
+		d.WriteString("api error: name = BadRequest" + " field = " + strings.Join(f, " ") +
+			" desc = " + strings.Join(desc, " ") + "\n")
+	}
+
+	if a.details.PreconditionFailure != nil {
+		v := a.details.PreconditionFailure.GetViolations()
+		var t []string
+		var s []string
+		var desc []string
+		for _, x := range v {
+			t = append(t, x.Type)
+			s = append(s, x.Subject)
+			desc = append(desc, x.Description)
+		}
+		d.WriteString("api error: name = PreconditionFailure" + " type = " + strings.Join(t, " ") +
+			" subj = " + strings.Join(s, " ") + " desc = " + strings.Join(desc, " ") + "\n")
+	}
+
+	if a.details.QuotaFailure != nil {
+		v := a.details.QuotaFailure.GetViolations()
+		var s []string
+		var desc []string
+		for _, x := range v {
+			s = append(s, x.Subject)
+			desc = append(desc, x.Description)
+		}
+		d.WriteString("api error: name = QuotaFailure" + " subj = " + strings.Join(s, " ") +
+			" desc = " + strings.Join(desc, " ") + "\n")
+	}
+
+	if a.details.RequestInfo != nil {
+		d.WriteString("api error: name = RequestInfo" + " id = " + a.details.RequestInfo.RequestId +
+			" data = " + a.details.RequestInfo.ServingData + "\n")
+	}
+
+	if a.details.ResourceInfo != nil {
+		d.WriteString("api error: name = ResourceInfo" + " type = " + a.details.ResourceInfo.ResourceType +
+			" resourcename = " + a.details.ResourceInfo.ResourceName + " owner = " + a.details.ResourceInfo.Owner +
+			" desc = " + a.details.ResourceInfo.Description + "\n")
+
+	}
+	if a.details.RetryInfo != nil {
+		d.WriteString("api error: name = RetryInfo" + " seconds = " + strconv.Itoa(int(a.details.RetryInfo.RetryDelay.Seconds)) +
+			"\n")
+
+	}
+	if a.details.Unknown != nil {
+		var s []string
+		for _, x := range a.details.Unknown {
+			s = append(s, fmt.Sprintf("%v", x))
+		}
+		d.WriteString("api error: name = Unknown" + " desc = " + strings.Join(s, " ") + "\n")
+
+	}
+	if a.details.DebugInfo != nil {
+		stack := strings.Join(a.details.DebugInfo.StackEntries, " ")
+		d.WriteString("api error: name = DebugInfo" + " detail = " + a.details.DebugInfo.Detail + " stack = " + stack + "\n")
+	}
+	if a.details.Help != nil {
+		var desc string
+		var url string
+		for _, x := range a.details.Help.Links {
+			desc = x.Description
+			url = x.Url
+		}
+		d.WriteString("api error: name = Help" + " desc = " + desc + " url = " + url + "\n")
+	}
+	if a.details.LocalizedMessage != nil {
+		d.WriteString("api error: name = LocalizedMessge" + " locale = " + a.details.LocalizedMessage.Locale +
+			" msg = " + a.details.LocalizedMessage.Message + "\n")
+	}
+	return d.String()
+}
+
+//GRPCStatus extracts underlying gRPC status in APIError
+func (a *APIError) GRPCStatus() *status.Status {
+	return a.status
+}
+
+//FromError extracts gRPC status from an error and builds an APIError
+func FromError(err error) (*APIError, bool) {
+	if err == nil {
+		return nil, false
+	}
+	msg := ErrDetails{}
```

Comment:
please move this initialization down to just before the `for` loop. The allocation is unnecessary if `status.FromError` is `!ok`.